### ISSUE TITLE
Add ability to export / list dependency trees/graphs

### ISSUE THREAD

**Issue by jamietanna** [renovatebot/renovate#18703]
### What would you like Renovate to be able to do?

Currently, there's no way to export the list of dependencies that are determined for a given repository.

This would make it possible to perform further analysis on dependencies, while taking advantage of Renovate's excellent suite of supported package managers.

### If you have any ideas on how this should be implemented, please tell us here.

As discussed in https://github.com/renovatebot/renovate/discussions/13150 there are some ways to hack it out of the ?. As [blogged about](https://www.jvt.me/posts/2022/11/01/renovate-dependency-graph/), it's [possible to use](https://gitlab.com/tanna.dev/renovate-graph/blob/89956637fa6babcc0f9750bf419e3276a04a91c7/index.ts#L104) the `PackageFiles`'s internal `data` map, but it's [a `private static`](https://github.com/renovatebot/renovate/blob/231d0a83258cb97db5ad4bf93fb61a9e7585e01f/lib/workers/repository/package-files.ts#L6-L7) so shouldn't really be depended on

We could:

- provide a CLI flag
- provide a separate CLI that can be distributed for this
- provide a straightforward TypeScript interface to glean the data, without needing to understand the internals of how the configuration is set up internally, processing secrets, etc

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by PhilipAbed** [comment]
i don't think this is in renovate's scope, 
@rarkins  @viceice  please take a look

**Comment by jamietanna** [comment]
Looks like we can actually remove the dependency on `PackageFiles`, as when we're using `dryRun: extract`, we get back the information we need. Don't know why I ended up going down the route of `PackageFiles` in the end, but I'm amending my tool to remove that need.

**Comment by PhilipAbed** [comment]
i myself worked on getting tree's for  almost all package managers out there, it is different for each, you will need a lot of work on each package manager to get 
1) all direct dependencies ( which we already have)
2) all dependencies of each package for all packages which will be recursive since u need to get children of children etc.
i certainly don't think this is in renovate's scope, but if the maintainers agree to add it, be my guest.

**Comment by viceice** [comment]
indirect deps will only possible to know if we have a lockfile to read them from.

resolving them from a remote registry would be too expensive for renovate.

**Comment by jamietanna** [comment]
Yes I think that's fair - to reword my ask, I think asking for the whole _graph_ / _tree_ may be a bit unfair to ask for. But knowing a list of:

- all direct
- all indirect/transitive dependencies

Would be handy.

In light of the indirects not being (easily) available, making it straightforward to export just the direct dependencies would be handy, but not sure if we think enough people need it to add on the maintenance of an easy entrypoint for TypeScript to do this?

**Comment by PhilipAbed** [comment]
Notes: 
npm version 7+ have extra information in package-lock.json
yarn 1 and yarn 2 lockfiles also have difference in their syntax
for every package manager there should be handlers for each type of package file,
which manager do you want to start this for?

**Comment by viceice** [comment]
correctly parsing the lockfiles would also probably help in our transitive remediation feature if we can parse them to a general structure. 🤔

**Comment by viceice** [comment]
dotnet lockfiles also have simple lockfiles, where all deps are marked with type (direct, transitive, ...)

**Comment by rarkins** [comment]
When lock file are present, we already parse them in order to determine the `lockedVersion` value. It would be OK if the reset of the parsed information is stored/returned semantically for output, because it wouldn't add much complexity/load.
